packages/java/parser-jvm-core/src/main/java/dev/hilla/parser/core/Plugin.java

```diff
@@ -1,32 +1,46 @@
 package dev.hilla.parser.core;
 
 import java.util.Collection;
+import java.util.Collections;
 
 import javax.annotation.Nonnull;
 
-import dev.hilla.parser.models.ClassInfoModel;
+public interface Plugin extends Comparable<Plugin> {
+    void enter(NodePath<?> nodePath);
 
-public interface Plugin {
-    int getOrder();
+    void exit(NodePath<?> nodePath);
 
-    void setOrder(int order);
+    PluginConfiguration getConfiguration();
 
-    default void setConfig(PluginConfiguration config) {
-        if (config != null) {
-            throw new IllegalArgumentException(String.format(
-                    "The '%s' plugin does not expect configuration set",
-                    getClass().getName()));
-        }
-    }
+    void setConfiguration(PluginConfiguration configuration);
 
-    void setStorage(@Nonnull SharedStorage storage);
+    default Collection<Class<? extends Plugin>> runAfter() {
+        return Collections.emptyList();
+    }
 
-    interface Preprocessor extends Plugin {
-        void preprocess();
+    default Collection<Class<? extends Plugin>> runBefore() {
+        return Collections.emptyList();
     }
 
-    interface Processor extends Plugin {
-        void process(@Nonnull Collection<ClassInfoModel> endpoints,
-                @Nonnull Collection<ClassInfoModel> entities);
+    @Nonnull
+    NodeDependencies scan(@Nonnull NodeDependencies nodeDependencies);
+
+    void setStorage(SharedStorage storage);
+
+    @Override
+    default int compareTo(Plugin o) {
+        if (runAfter().contains(o.getClass())) {
+            return 1;
+        }
+        if (runBefore().contains(o.getClass())) {
+            return -1;
+        }
+        if (o.runAfter().contains(getClass())) {
+            return -1;
+        }
+        if (o.runBefore().contains(getClass())) {
+            return 1;
+        }
+        return getClass().getName().compareTo(o.getClass().getName());
```

Comment:
This should allow to test for invalid configurations (would need to add a test for that):

```suggestion
        int result = runBefore().contains(other.getClass())
                || other.runAfter().contains(getClass()) ? -1 : 0;
        if (runAfter().contains(other.getClass())
                || other.runBefore().contains(getClass())) {
            result ^= 1;
        }
        if (result == -2) {
            throw new IllegalStateException(String.format(
                    "Plugins %s and %s have conflicting order needs",
                    getClass().getName(), other.getClass().getName()));
        }
        return result;
```